google-cloud-firestore/src/main/java/com/google/cloud/firestore/FieldPath.java

```diff
@@ -152,4 +152,9 @@ FieldPath createPathWithSegments(ImmutableList<String> segments) {
   public String toString() {
     return getEncodedPath();
   }
+
+  @Override
+  public int compareTo(@Nonnull FieldPath other) {
+    return canonicalString().compareTo(other.canonicalString());
```

Comment:
This comparison seems expensive. I think we could do something like this? https://play.golang.org/p/-iP464mWSE .

Sorry for using Go here. I use it mostly for testing the code quickly.

Reply:
I changed it to use the cached version. I agreed that it is not the most performance efficient, but
- the input is expected to be small (as it is the list of field path)
- we have to compute the canonical String anyways and this now (with the recent change) already populates the cache
- It's super simple and works since it's already escaped (otherwise we have to compare every element to its corresponding value like you did in your Go code).